### PR TITLE
AEA-851 MultiplexerStatus

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

Improvement of the multiplexer status object

## Fixes

https://github.com/fetchai/agents-aea/issues/1689

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/multiplexer.py [hunk 1]
```diff
@@ -201,30 +218,36 @@ async def connect(self) -> None:
...
                 return
             try:
+                self.connection_status.set(ConnectionStates.connecting)
                 await self._connect_all()
-                enforce(self.is_connected, "At least one connection failed to connect!")
-                self._connection_status.is_connected = True
+
+                if all(c.is_connected for c in self._connections):
+                    self.connection_status.set(ConnectionStates.connected)
+                elif not all(
```
Comment (DavidMinarsch) [inline aea/multiplexer.py:226]: why not simply `else`? If not all connections are connected then it means `connect` of at least one connection has been implemented wrongly or failed so we should raise?

File: aea/multiplexer.py [hunk 2]
```diff
@@ -201,30 +218,36 @@ async def connect(self) -> None:
...
         async with self._lock:
-            if not self.connection_status.is_connected:
+            if self.connection_status.is_disconnected:
                 self.logger.debug("Multiplexer already disconnected.")
                 await asyncio.wait_for(self._stop(), timeout=60)
                 return
             try:
+                self.connection_status.set(ConnectionStates.disconnecting)
                 await asyncio.wait_for(self._disconnect_all(), timeout=60)
                 await asyncio.wait_for(self._stop(), timeout=60)
```
Comment (DavidMinarsch) [inline aea/multiplexer.py:250]: at the end we should check:
```
if all(c.is_disconnected for c in self._connections):
       self.connection_status.set(ConnectionStates.disconnected)
else:
     raise ...
```

File: aea/multiplexer.py [hunk 3]
```diff
@@ -253,6 +276,10 @@ async def _stop(self) -> None:
             if c.state in (ConnectionStates.connecting, ConnectionStates.connected)
         ]:
             await connection.disconnect()
+
```
Comment (DavidMinarsch) [inline aea/multiplexer.py:279]: I am confused how this related to `disconnect` above. Should we not be calling `disconnect` from here and implement the disconnection in `disconnect`?

Comment (solarw) [reply aea/multiplexer.py:279]: updated.

File: aea/multiplexer.py (hunk 1 above)
Comment (solarw) [reply aea/multiplexer.py:226]: updated, but should discuss

File: aea/multiplexer.py (hunk 2 above)
Comment (solarw) [reply aea/multiplexer.py:250]: updated, but should discuss

File: aea/multiplexer.py [hunk 4]
```diff
@@ -40,17 +40,34 @@
...
     def __init__(self):
         """Initialize the connection status."""
-        self.is_connected = False  # type: bool
-        self.is_connecting = False  # type: bool
+        super().__init__(
+            initial_state=ConnectionStates.disconnected, states_enum=ConnectionStates
+        )
+
+    @property
+    def is_connected(self) -> bool:  # pragma: nocover
```
Comment (DavidMinarsch) [inline aea/multiplexer.py:53]: I think it is worth testing these properties...

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**